cpp/src/glm/qn/qn_solvers.cuh

```diff
@@ -139,10 +139,16 @@ inline OPT_RETCODE min_lbfgs(const LBFGSParam<T> &param,
     LINE_SEARCH_RETCODE lsret =
       ls_backtrack(param, f, fx, x, grad, step, drt, xp, dev_scalar, stream);
 
-    bool isLsSuccess = lsret == LS_SUCCESS;
-    CUML_LOG_TRACE("Iteration %d, fx=%f", *k, fx);
-
-    if (!isLsSuccess || isnan(fx) || isinf(fx)) {
+    bool isLsValid = !isnan(fx) && !isinf(fx);
+    // Linesearch may fail to converge, but still come closer to the solution;
+    // if that is not the case, let `check_convergence` ("insufficient change")
+    // below terminate the loop.
+    bool isLsInDoubt =
+      lsret == LS_INVALID_STEP_MIN || lsret == LS_MAX_ITERS_REACHED;
+    bool isLsSuccess =
+      lsret == LS_SUCCESS || isLsValid && fx <= fxp && isLsSuccess;
```

Comment:
Does this last condition have the intended affect? It appears to be using `isLsSuccess` in the initial assignment of `isLsSuccess`. Does `isLsInDoubt` get used anywhere?

Reply:
Oops, that's embarrassing :)